File: core/src/project_table_store.cpp
#include "streamfind/project_table_store.hpp"
#include "streamfind/catalogue.hpp"

#include <algorithm>
#include <sstream>

namespace streamfind {

ProjectTableStore::ProjectTableStore(Project &project) noexcept : project_(&project) {}

bool ProjectTableStore::has_table(const std::string &table_name) const {
    const auto tables = project_->list_tables();
    return std::find(tables.begin(), tables.end(), table_name) != tables.end();
}

void ProjectTableStore::require(const std::vector<TableRequirement> &requirements) const {
    for (const auto &requirement : requirements) {
        if (!has_table(requirement.name))
            throw Error(ErrorCode::SchemaMismatch, "missing domain table: " + requirement.name);
        if (requirement.required_columns.empty() && requirement.required_column_types.empty()) continue;
        const auto rows = project_->query_json("DESCRIBE \"" + requirement.name + "\"");
        for (const auto &column : requirement.required_columns) {
            const auto found = std::find_if(rows.begin(), rows.end(), [&column](const auto &row) {
                return row.value("column_name", "") == column;
            });
            if (found == rows.end())
                throw Error(ErrorCode::SchemaMismatch,
                            "missing column " + column + " in domain table " + requirement.name);
        }
        for (const auto &[column, semantic_type] : requirement.required_column_types) {
            const auto found = std::find_if(rows.begin(), rows.end(), [&column](const auto &row) {
                return row.value("column_name", "") == column;
            });
            if (found == rows.end())
                throw Error(ErrorCode::SchemaMismatch,
                            "missing column " + column + " in domain table " + requirement.name);
            const std::string actual = found->value("column_type", "");
            const auto hash = semantic_type.rfind('#');
            const std::string expected = hash == std::string::npos ? semantic_type : semantic_type.substr(hash + 1);
            const bool compatible = expected == "string" ? (actual == "VARCHAR" || actual == "TEXT")
                : expected == "integer" ? actual.find("INT") != std::string::npos
                : expected == "real" ? (actual == "DOUBLE" || actual == "FLOAT" || actual == "DECIMAL")
                : expected == "boolean" ? actual == "BOOLEAN"
                : expected == "timestamp" ? actual.find("TIMESTAMP") != std::string::npos
                : true;
            if (!compatible)
                throw Error(ErrorCode::SchemaMismatch, "incompatible type for " + column + " in domain table " +
                                                        requirement.name + ": expected " + expected + ", got " + actual);
        }
    }
}

void ProjectTableStore::require_manifest(const std::string &domain) const {
    const auto manifest = catalogue::table_manifest_json(domain);
    if (!manifest) throw Error(ErrorCode::DatabaseError, "generated table manifest unavailable");
    std::vector<TableRequirement> requirements;
    for (const auto &table : *manifest) {
        TableRequirement requirement;
        requirement.name = table.value("table_name", "");
        for (const auto &column : table.value("columns", Json::array()))
            requirement.required_column_types.emplace_back(column.value("name", ""), column.value("type", ""));
        requirements.push_back(std::move(requirement));
    }
    require(requirements);
}

void ProjectTableStore::ensure_table(const std::string &, const std::string &ddl) const {
    project_->execute_sql(ddl);
}

Json ProjectTableStore::query(const std::string &sql) const { return project_->query_json(sql); }

void ProjectTableStore::execute(const std::string &sql) const { project_->execute_sql(sql); }

void ProjectTableStore::append(
    const std::string &table_name, const std::vector<std::string> &column_names,
    const std::vector<std::vector<std::optional<std::string>>> &rows) const {
    project_->append_rows(table_name, column_names, rows);
}

}  // namespace streamfind

File: rust/crates/core/src/project_table_store.rs
use crate::{Error, ErrorCode, Json, Project, Result};

/// A table and the columns a domain module requires before execution.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TableRequirement {
    pub name: String,
    pub required_columns: Vec<String>,
}

/// Narrow table-access boundary exposed to domain modules.
pub struct ProjectTableStore<'a> {
    project: &'a Project,
}

impl<'a> ProjectTableStore<'a> {
    pub fn new(project: &'a Project) -> Self {
        Self { project }
    }

    pub fn has_table(&self, table_name: &str) -> Result<bool> {
        Ok(self
            .project
            .list_tables()?
            .iter()
            .any(|name| name == table_name))
    }

    pub fn require(&self, requirements: &[TableRequirement]) -> Result<()> {
        for requirement in requirements {
            if !self.has_table(&requirement.name)? {
                return Err(Error::new(
                    ErrorCode::SchemaMismatch,
                    format!("missing domain table: {}", requirement.name),
                ));
            }
            if requirement.required_columns.is_empty() {
                continue;
            }
            let rows = self
                .project
                .query_json(&format!("DESCRIBE \"{}\"", requirement.name))?;
            for column in &requirement.required_columns {
                if !rows
                    .as_array()
                    .into_iter()
                    .flatten()
                    .any(|row| row["column_name"].as_str() == Some(column))
                {
                    return Err(Error::new(
                        ErrorCode::SchemaMismatch,
                        format!(
                            "missing column {column} in domain table {}",
                            requirement.name
                        ),
                    ));
                }
            }
        }
        Ok(())
    }

    /// Validate every table contract generated from the semantic catalogue.
    pub fn require_manifest(&self, domain: &str) -> Result<()> {
        let manifest = crate::catalogue::table_manifest(domain)
            .map_err(|error| Error::new(ErrorCode::SchemaMismatch, error.to_string()))?;
        for (table_name, columns) in manifest {
            if !self.has_table(&table_name)? {
                return Err(Error::new(
                    ErrorCode::SchemaMismatch,
                    format!("missing domain table: {table_name}"),
                ));
            }
            let rows = self
                .project
                .query_json(&format!("DESCRIBE \"{}\"", table_name))?;
            for (column_name, semantic_type) in columns {
                let Some(row) = rows
                    .as_array()
                    .into_iter()
                    .flatten()
                    .find(|row| row["column_name"].as_str() == Some(column_name.as_str()))
                else {
                    return Err(Error::new(
                        ErrorCode::SchemaMismatch,
                        format!("missing column {column_name} in domain table {table_name}"),
                    ));
                };
                let expected = semantic_type.rsplit('#').next().unwrap_or(&semantic_type);
                let actual = row["column_type"]
                    .as_str()
                    .unwrap_or_default()
                    .to_ascii_uppercase();
                let compatible = match expected {
                    "boolean" => actual == "BOOLEAN",
                    "integer" => actual.contains("INT"),
                    "real" => actual == "DOUBLE" || actual == "FLOAT" || actual == "DECIMAL",
                    "string" => actual == "VARCHAR" || actual == "TEXT",
                    "timestamp" => actual.contains("TIMESTAMP"),
                    _ => true,
                };
                if !compatible {
                    return Err(Error::new(
                        ErrorCode::SchemaMismatch,
                        format!("incompatible type for {column_name} in {table_name}: expected {expected}, got {actual}"),
                    ));
                }
            }
        }
        Ok(())
    }

    pub fn ensure_table(&self, ddl: &str) -> Result<()> {
        self.project.execute_sql(ddl)
    }

    pub fn execute(&self, sql: &str) -> Result<()> {
        self.project.execute_sql(sql)
    }

    pub fn query(&self, sql: &str) -> Result<Json> {
        self.project.query_json(sql)
    }
}
